### PR TITLE
kernel: Rewrite resource limit to be more accurate

### PR DESCRIPTION
**Pull request by ogniK5377**
Matches closer to hardware

### CODE REVIEW

File: src/core/hle/kernel/process.cpp [hunk 1]
```diff
@@ -116,7 +116,7 @@ std::shared_ptr<Process> Process::Create(Core::System& system, std::string name,
 
     std::shared_ptr<Process> process = std::make_shared<Process>(system);
     process->name = std::move(name);
-    process->resource_limit = ResourceLimit::Create(kernel);
+    process->resource_limit = std::make_shared<KResourceLimit>(kernel, system);
```
Comment (ameerj) [inline src/core/hle/kernel/process.cpp:119]: For hardware accuracy, the `Process` constructor holds a reference to the global kernel resource limit, not create a new one on the fly.

File: src/core/hle/kernel/k_light_condition_variable.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    void Wait(KLightLock* lock, s64 timeout = -1ll) {
+        WaitImpl(lock, timeout);
+        lock->Lock();
+    }
+
+    void Broadcast() {
+        KScopedSchedulerLock lk{kernel};
+        while (m_thread_queue.WakeupFrontThread() != nullptr) {
+            /* We want to signal all threads, and so should continue waking up until there's nothing
```
Comment (lioncash) [inline src/core/hle/kernel/k_light_condition_variable.h:34]: We use `//`-style comments, Ditto elsewhere below

File: src/core/hle/kernel/k_light_condition_variable.h [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+class KernelCore;
+
+class KLightConditionVariable {
+private:
+    KThreadQueue m_thread_queue;
+
+public:
+    KLightConditionVariable(KernelCore& kernel) : m_thread_queue(kernel), kernel(kernel) {}
+
+    void Wait(KLightLock* lock, s64 timeout = -1ll) {
```
Comment (lioncash) [inline src/core/hle/kernel/k_light_condition_variable.h:26]:
```suggestion
    void Wait(KLightLock* lock, s64 timeout = -1) {
```

File: src/core/hle/kernel/k_light_condition_variable.h [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+namespace Kernel {
+class KernelCore;
+
+class KLightConditionVariable {
+private:
+    KThreadQueue m_thread_queue;
+
+public:
+    KLightConditionVariable(KernelCore& kernel) : m_thread_queue(kernel), kernel(kernel) {}
```
Comment (lioncash) [inline src/core/hle/kernel/k_light_condition_variable.h:24]:
```suggestion
    explicit KLightConditionVariable(KernelCore& kernel) : m_thread_queue(kernel), kernel(kernel) {}
```

File: src/core/hle/kernel/k_light_condition_variable.h [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+#include "core/hle/kernel/k_scheduler.h"
+#include "core/hle/kernel/k_scoped_scheduler_lock_and_sleep.h"
+#include "core/hle/kernel/k_thread_queue.h"
+#include "core/hle/kernel/time_manager.h"
+
+namespace Kernel {
+class KernelCore;
+
+class KLightConditionVariable {
+private:
```
Comment (lioncash) [inline src/core/hle/kernel/k_light_condition_variable.h:20]: This should be moved after the public interface

File: src/core/hle/kernel/k_light_condition_variable.h [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        while (m_thread_queue.WakeupFrontThread() != nullptr) {
+            /* We want to signal all threads, and so should continue waking up until there's nothing
+             * to wake. */
+        }
+    }
+
+private:
+    void WaitImpl(KLightLock* lock, s64 timeout) {
+        KThread* owner = GetCurrentThreadPointer(kernel);
+        // KHardwareTimer* timer;
```
Comment (lioncash) [inline src/core/hle/kernel/k_light_condition_variable.h:42]: Leftover code?

File: src/core/hle/kernel/k_resource_limit.cpp [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+#include "common/assert.h"
+#include "core/core.h"
+#include "core/core_timing.h"
+#include "core/core_timing_util.h"
+#include "core/hle/kernel/k_resource_limit.h"
+#include "core/hle/kernel/svc_results.h"
+
+namespace Kernel {
+namespace {
+constexpr s64 DefaultTimeout = 10000000000; // 10 seconds
```
Comment (lioncash) [inline src/core/hle/kernel/k_resource_limit.cpp:17]: This doesn't need to be enclosed in an anonymous namespace, given its already internally linked by virtue of being constexpr

File: src/core/hle/kernel/k_resource_limit.h [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    Count,
+};
+
+constexpr bool IsValidResourceType(LimitableResource type) {
+    return type < LimitableResource::Count;
+}
+
+class KResourceLimit final : public Object {
+public:
+    KResourceLimit(KernelCore& kernel, Core::System& system);
```
Comment (lioncash) [inline src/core/hle/kernel/k_resource_limit.h:40]:
```suggestion
    explicit KResourceLimit(KernelCore& kernel, Core::System& system);
```

File: src/core/hle/kernel/k_resource_limit.h [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+    static constexpr HandleType HANDLE_TYPE = HandleType::ResourceLimit;
+    HandleType GetHandleType() const override {
+        return HANDLE_TYPE;
+    }
+
+    virtual void Finalize() override {}
+
+private:
+    std::array<s64, static_cast<std::size_t>(LimitableResource::Count)> limit_values{};
```
Comment (lioncash) [inline src/core/hle/kernel/k_resource_limit.h:70]:
```cpp
using ResourceArray = std::array<s64, static_cast<std::size_t>(LimitableResource::Count)>;
```

and then reuse it for all the arrays, to reduce verbosity a little bit.

File: src/core/hle/kernel/k_light_condition_variable.h [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+#include "core/hle/kernel/k_scoped_scheduler_lock_and_sleep.h"
+#include "core/hle/kernel/k_thread_queue.h"
+#include "core/hle/kernel/time_manager.h"
+
+namespace Kernel {
+class KernelCore;
+
+class KLightConditionVariable {
+private:
+    KThreadQueue m_thread_queue;
```
Comment (lioncash) [inline src/core/hle/kernel/k_light_condition_variable.h:21]: Is the `m_` prefix meant to be here? It was dropped off another variable previously in another commit

File: src/core/hle/kernel/k_resource_limit.h [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+namespace Core {
+class System;
+}
+
+namespace Kernel {
+class KernelCore;
+enum class LimitableResource : u32 {
+    PhysicalMemoryMax = 0,
+    ThreadCountMax = 1,
```
Comment (lioncash) [inline src/core/hle/kernel/k_resource_limit.h:26]: Maybe consider dropping the Max from the names or something. `LimitableResource` is already a lot to type, so

```cpp
LimitableResource::Threads
```

or

```cpp
LimitableResource::MaxThreads
```

is much more concise than

```cpp
LimitableResource::ThreadCountMax
```

Ditto for the other entries in the enum

**Review by bunnei — APPROVED** [review]

**Pull request merged**